# 0. Contents

1. products consistency check

2 . orders consistency checks 

# 1. Importing Data

In [53]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [54]:
# defining main path
path = r'C:\Users\muffi\OneDrive\CareerFoundry\Instacart Basket Analysis 2023'

In [55]:
# importing OG products dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [56]:
# importing wrangled orders dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled3.csv'), index_col = False)

# 2. Data Consistency Checks

Mixed-data types should always be checked for since they can break your code

Make sure to keep prep scripts separate from analysis scripts

## 2.1. Mixed-Type Data

In [57]:
# creating practice dataset
df_test = pd.DataFrame()

In [58]:
# adding a mixed col
df_test['mix'] = ['a', 'b', 1, True]

In [59]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [60]:
# checking manually for mixed-type cols

# use a for-loop to iterate through the cols
# weirld is a boolean that checks if the current col has inconsistent data types
# if weird > 1, then weird = True
# If weird is true for the col, then print the col

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [61]:
# changing data type
df_test['mix'] = df_test['mix'].astype('str')

In [62]:
# checking dataframe
#df_test['mix'].dtype

## 2.2. Missing Values

In [63]:
# checking for missing data

# .isnull() tells whether a cell is missing data or not
# .sum() adds together the number missing in each row (because True is 1 and False is 0)

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [64]:
# making a dataframe with the null rows
df_nan = df_prods.loc[df_prods['product_name'].isnull() == True]

In [65]:
# checking the null dataframe
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Recap of ways to deal with missing data:
    1. Create a new variable that flags missing values
    2. Impute mean or median (for numeric values)
    3. Remove or filter missing data

In [66]:
# replacing missing vals with the mean
#df['col with missing vals'].fillna(mean value, inplace = True)

# replacing missing cals with the median
#df[' col with missing vals'].fillna(median value, inplace = True)

In [67]:
# seeing how many rows I have originally
df_prods.shape

(49693, 5)

In [68]:
# creating a clean dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [69]:
# checking the size of the cleaned dataframe
df_prods_clean.shape

(49677, 5)

In [70]:
# another way to drop NaNs
#df_prods.dropna(inplace = True)

# dropping NaNs from a specific col
#df_prods.dropna(subset = ['product_name'], inpace = True)

## 2.3. Duplicates

When dealing with duplicates, it is importnt to record occurences, export examples, and send them to the client for futher explanation


In [71]:
# finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [72]:
# checking dups
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [73]:
# checking current number of rows
df_prods_clean.shape

(49677, 5)

In [74]:
df_prods_clean_dupless = df_prods_clean.drop_duplicates()

In [75]:
df_prods_clean_dupless.shape

(49672, 5)

## 2.4. Exporting Checked Data

In [76]:
# exporing checked products dataframe
df_prods_clean_dupless.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 3. Exercise Script

## 3.1. Orders Descriptive Statistics

In [77]:
# running descriptive analysis
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

order_id and user_id are int and they should be strings

orders_day_of_week looks good

order_hour_of_day looks good

days_since_prior_order is missing data

order_number has a max of 100 and a q3 of 23 (max seems high)

## 3.2. Checking for Mixed-type Data

In [78]:
# checking for mixed data types
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


There is no mixed-type data

## 3.3. Checking for Missing Values

In [79]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 cells in days_since_prior order that are missing values. These are probably first order, thus there wouldn't be an order prior. 

## 3.4. Addressing Missing Values

In [80]:
# checking orginal dataframe dimensions
df_ords.shape

(3421083, 7)

In [81]:
# checking % of missing data
# if it is greater than 5% then dropping would not be a good choice
206209/3421083

0.06027594185817766

In [82]:
# creating a table of just the nulls
df_ords_nulls = df_ords.loc[df_ords['days_since_prior_order'].isnull() == True]

In [83]:
# checking the nulls dataframe
df_ords_nulls.tail(20)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
3420678     3420678   1955107   206190             1                   6   
3420687     3420687    936117   206191             1                   6   
3420693     3420693   3242108   206192             1                   6   
3420708     3420708    313609   206193             1                   5   
3420750     3420750   3395125   206194             1                   1   
3420762     3420762    926652   206195             1                   3   
3420782     3420782    825502   206196             1                   6   
3420787     3420787   3325493   206197             1                   2   
3420811     3420811   2154947   206198             1                   0   
3420819     3420819   3302466   206199             1                   2   
3420839     3420839   2508134   206200             1                   5   
3420863     3420863     28210   206201             1                   6   
3420896     3420896    799173   206202             1                   2   
3420919     3420919   1991850   206203             1                   1   
3420925     3420925   1438269   206204             1                   1   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
3420678                  9                     NaN  
3420687                 15                     NaN  
3420693                  6                     NaN  
3420708                 15                     NaN  
3420750                  9                     NaN  
3420762                 16                     NaN  
3420782                 14                     NaN  
3420787                 19                     NaN  
3420811                 16                     NaN  
3420819                 11                     NaN  
3420839                 19                     NaN  
3420863                 14                     NaN  
3420896                 12                     NaN  
3420919                 13                     NaN  
3420925                 11                     NaN  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN

In [84]:
# checking to see if imputting 0 is reasonable 
df_ords[df_ords['days_since_prior_order'] == 0]

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
8                 8   2295261        1             9                   1   
43               43   2557754        4             5                   5   
189             189   2148135       17            30                   6   
296             296    965160       24            19                   0   
337             337   3162630       27            24                   2   
...             ...       ...      ...           ...                 ...   
3420980     3420980      4277   206206            47                   2   
3420985     3420985    648225   206206            52                   0   
3420991     3420991   2475308   206206            58                   0   
3421001     3421001   1043943   206206            68                   0   
3421044     3421044   2079142   206208            26                   2   

         order_hour_of_day  days_since_prior_order  
8                       16                     0.0  
43                      13                     0.0  
189                     15                     0.0  
296                     16                     0.0  
337                     15                     0.0  
...                    ...                     ...  
3420980                 17                     0.0  
3420985                 19                     0.0  
3420991                 15                     0.0  
3421001                 20                     0.0  
3421044                 16                     0.0  

[67755 rows x 7 columns]

In [87]:
# replacing missing values with 0s 
#df['column with missings'].fillna(median value, inplace=True)
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [88]:
# checking for nulls in the cleaned dataframe
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

Since more than 6% of the data was missing, and the null cell contained valuable information about first orders, I decided not to drop the values. Instead, I chose to imput 0 for those values, because it was techinically true, and would make it easy to find. The only hesitation I had was that 67,755 rows that were ordered in the same day as their previous order that would have a 0 in the days_since_prior_order columns. This would be fine, because the two type can be distingusihed using the order_number column (first orders would have the number 1 in this column)

## 3.5. Checking for Duplicates

In [90]:
# checking for dulplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [94]:
# seeing the duplicates
df_ords_dups.shape

(0, 7)

No duplicates were found. Seems like everyone did a great job collecting data. 

In [96]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))